openedx/core/djangoapps/programs/utils.py

```diff
@@ -341,34 +343,52 @@ def supplement_program_data(program_data, user):
 
-            run_mode['course_url'] = reverse('course_root', args=[course_key])
-            run_mode['course_image_url'] = course_overview.course_image_url
+            course_url = reverse('course_root', args=[course_key])
+            course_image_url = course_overview.course_image_url
 
-            run_mode['start_date'] = course_overview.start_datetime_text()
-            run_mode['end_date'] = course_overview.end_datetime_text()
+            start_date_string = course_overview.start_datetime_text()
+            end_date_string = course_overview.end_datetime_text()
 
             end_date = course_overview.end or datetime.datetime.max.replace(tzinfo=pytz.UTC)
-            run_mode['is_course_ended'] = end_date < timezone.now()
+            is_course_ended = end_date < timezone.now()
 
-            run_mode['is_enrolled'] = CourseEnrollment.is_enrolled(user, course_key)
+            is_enrolled = CourseEnrollment.is_enrolled(user, course_key)
 
             enrollment_start = course_overview.enrollment_start or datetime.datetime.min.replace(tzinfo=pytz.UTC)
             enrollment_end = course_overview.enrollment_end or datetime.datetime.max.replace(tzinfo=pytz.UTC)
             is_enrollment_open = enrollment_start <= timezone.now() < enrollment_end
-            run_mode['is_enrollment_open'] = is_enrollment_open
-            if not is_enrollment_open:
-                # Only render this enrollment open date if the enrollment open is in the future
-                run_mode['enrollment_open_date'] = strftime_localized(enrollment_start, 'SHORT_DATE')
 
-            # TODO: Currently unavailable on LMS.
-            run_mode['marketing_url'] = ''
+            enrollment_open_date = None if is_enrollment_open else strftime_localized(enrollment_start, 'SHORT_DATE')
 
             certificate_data = certificate_api.certificate_downloadable_status(user, course_key)
             certificate_uuid = certificate_data.get('uuid')
-            if certificate_uuid:
-                run_mode['certificate_url'] = certificate_api.get_certificate_url(
-                    course_id=course_key,
-                    uuid=certificate_uuid,
-                )
+            certificate_url = certificate_api.get_certificate_url(
+                course_id=course_key,
+                uuid=certificate_uuid,
+            ) if certificate_uuid else None
+
+            required_mode_slug = run_mode['mode_slug']
+            enrolled_mode_slug, _ = CourseEnrollment.enrollment_mode_for_user(user, course_key)
+            is_mode_mismatch = required_mode_slug != enrolled_mode_slug
+            is_upgrade_required = is_enrolled and is_mode_mismatch
+
+            # Assumes that the ecommerce service is in use.
+            required_mode = CourseMode.mode_for_course(course_key, required_mode_slug)
+            upgrade_url = EcommerceService().checkout_page_url(required_mode.sku) if is_upgrade_required else None
```

Comment:
My concern is, if the setup do not have ecommerce IDA setup, would EcommerceService object below handle correctly and return either `None` or empty string?